src/deepdisc/data_format/file_io.py

```diff
@@ -1,5 +1,173 @@
 import json
 from pathlib import Path
+import glob
+from astropy.io import fits
+import os
+import cv2
+import numpy as np
+from detectron2.structures import BoxMode
+
+
+
+class DDLoader:
+    """A base deepdisc data loader class"""
+
+    def __init__(self):
+        self.filedict = None
+        self.dataset = None
+
+    def get_dataset(self):
+        """retrieves the list of dataset_dicts if established."""
+        return self.dataset
+
+    def generate_filedict(
+        self, dirpath, filters, img_files, mask_files, subdirs=False, filt_loc=0, n_samples=None
+    ):
+        """Generates a path dictionary from a directory of files.
+
+        Parameters
+        ----------
+        dirpath : str, path-like
+            The path to the data directory.
+        filters : list
+            A list of filters available in the dataset. The filter names should
+            match some string identifier in the name itself. E.g. img_r.fits
+            will be matched to a filter with label "r"
+        img_files: str
+            The name of the image files to collect, should have a "*" to
+            collect all image files in the dataset. E.g. 001_img.fits can be
+            caught used img_files = "*_img.fits"
+        maskfiles: str
+            The name of the mask files to collect, should have a "*" to
+            collect all mask files in the dataset. E.g. 001_mask.fits can be
+            caught used mask_files = "*_mask.fits"
+        subdirs: bool
+            Indicates whether the data is stored within subdirectories within
+            the dirpath. If True, will recursively search for files.
+        filt_loc: int
+            The integer location of the filter within the image name, used to
+            split files across filters accordingly. E.g. 001_img_r.fits would
+            have a filt_loc of 8 (or -6).
+        n_samples: int
+            If specified, filters down to a subset of the dataset that contains
+            `n_samples` image files per filter.
+
+        Returns
+        -------
+        self : DataLoader
+            A `DataLoader` with a filename dictionary generated in
+            `DataLoader.filedict`
+        """
+
+        # Setup filenames dict
+        filenames_dict = {}
+        filenames_dict["filters"] = filters
+
+        # Glob in filenames from the paths
+        if subdirs:
```

Comment:
Good question, I currently have it setup that those dataset directories should be part of the initial directory path, like so (from the hsc notebook): 

<img width="912" alt="Screen Shot 2023-11-14 at 2 30 31 PM" src="https://github.com/lincc-frameworks/deepdisc/assets/8853505/bb886a87-7f6d-41dc-8db3-62e172511f6d">

So the loader would be established for each of (train, test, val) separately, either by creating 3 loaders or doing some loop. My rationale for this was that it seemed like we wanted the various datasets fairly separate. Does that sound in line with what you'd want this to do?

Reply:
Ok, that makes sense and we can keep the philosophy of separate dataloaders for separate datasets.  When do you envision the subdirs option being used instead of supplying the full directory path?

Reply:
The main reason for subdirs was to add flexibility for differing directory structures. For HSC, all the data is right in the top level directory (.../HSC_DR3/data/train/[image files]), but for DECAM, the data is split among subdirectories per file (decam_data/train/set_xx/[image_files]). So for DECAM we would set the subdir flag to let it know to look within those subdirectories to get the files.

There may be a better way to do this?

Reply:
Ok, that makes sense.  I think that's good to keep in